lib/Sema/SemaBounds.cpp

```diff
@@ -578,6 +578,79 @@ namespace {
 
       return true;
     }
+
+
+    bool VisitImplicitCastExpr(ImplicitCastExpr *E) {
+      if (E->getCastKind() == CK_BitCast)
+        FindDisallowedFunctionPtrCasts(E);
+
+      return true;
+    }
+
+  private:
+    // Here we're looking for places where the programmer has cast from an
+    // unchecked function pointer type to a checked function pointer type.
+    // We need to descend through all the casts below the top-level bitcast to find
+    // the eventual thing we're referring to. We then check to make sure that it is
+    // either a top-level function declaration (of either checked or unchecked type), or
+    // a local variable that has a checked function type.
+    //
+    // The only other exception we have is that we can always assign the null pointer to
+    // a checked function pointer type.
+    void FindDisallowedFunctionPtrCasts(ImplicitCastExpr *E) {
+      QualType Ty = E->getType();
+
+      // We're looking for ptrs to functions
+      if (!Ty->isCheckedPointerPtrType() ||
+        !Ty->isFunctionPointerType())
+        return;
+
+      Expr *Operand = E;
+      // This is so we can get down through all & and * operators
+      // it is modelled on the Expr::IgnoreParenCasts-like functions
+      while (true) {
+        // This takes us down to the next "interesting" thing
+        Operand = Operand->IgnoreParenCasts();
```

Comment:
I think this code needs to be more restrictive.  I think an invalid function pointer value could be produced by a combination of explicit casts in the subexpression of the implicit cast,, LValueToRValue casts, and dereference operators.    OperationKinds.def in clang\include\clang\AST lists the different kinds of casts.